src/engine/IndexScan.cpp

```diff
@@ -324,15 +341,192 @@ Permutation::IdTableGenerator IndexScan::lazyScanForJoinOfColumnWithScan(
+  bool fetchFirst() {
+    AD_CORRECTNESS_CHECK(!iterator_.has_value());
+    iterator_ = generator_.begin();
+    while (iterator_ != generator_.end()) {
+      if (iterator_.value()->idTable_.empty()) {
+        ++iterator_.value();
+        continue;
+      }
+      hasUndef_ = iterator_.value()->idTable_.at(0, joinColumn_).isUndefined();
+      return false;
+    }
+    return true;
+  }
+
+  // Consume the next non-empty table from the generator and calculate the next
+  // matching blocks from the index scan.
+  void fetch() {
+    bool exhausted;
+    if (iterator_.has_value()) {
+      AD_CONTRACT_CHECK(iterator_.value() != generator_.end());
+      do {
+        // Skip empty tables
+        ++iterator_.value();
+        exhausted = iterator_.value() == generator_.end();
+      } while (!exhausted && iterator_.value()->idTable_.empty());
+    } else {
+      exhausted = fetchFirst();
+    }
+    if (!exhausted) {
+      prefetchedValues_.push_back(std::move(*iterator_.value()));
+      auto joinColumn =
+          prefetchedValues_.back().idTable_.getColumn(joinColumn_);
+      AD_EXPENSIVE_CHECK(std::ranges::is_sorted(joinColumn));
+      // Skip processing for undef case, it will be handled differently
+      if (joinColumn.empty() || !joinColumn[0].isUndefined()) {
+        AD_CORRECTNESS_CHECK(!hasUndef_);
+        // Find first value that differs from the last one that was used to find
+        // matching blocks.
+        size_t offset = std::ranges::find_if_not(joinColumn,
+                                                 [this](const auto& element) {
+                                                   return element == lastValue_;
+                                                 }) -
+                        joinColumn.begin();
+        auto newValues = joinColumn.subspan(offset);
+        if (!newValues.empty()) {
+          std::ranges::move(CompressedRelationReader::getBlocksForJoin(
+                                newValues, metaBlocks_),
+                            std::back_inserter(pendingBlocks_));
+          lastValue_ = newValues.back();
+        }
+      }
+    }
+    doneFetching_ = exhausted;
+  }
+
+  // Check if there are any undefined values yielded by the original generator.
+  // If the generator hasn't been started to get consumes, this will start it.
+  bool hasUndef() {
+    if (!iterator_.has_value() && !doneFetching_) {
+      fetch();
+    }
+    return hasUndef_;
+  }
+};
+
+// _____________________________________________________________________________
+Result::Generator IndexScan::createPrefilteredJoinSide(
+    std::shared_ptr<SharedGeneratorState> innerState) {
+  if (innerState->hasUndef()) {
+    for (Result::IdTableVocabPair& pair : innerState->prefetchedValues_) {
+      co_yield pair;
+    }
+    innerState->prefetchedValues_.clear();
+    auto& iterator = innerState->iterator_.value();
+    if (iterator != innerState->generator_.end()) {
+      ++iterator;
+    }
```

Comment:
These semantics are fishy and hard to reason about.
Can you make it, s.t. the iterator ALWAYS points to the first value that was NOT YET prefetched?